infinite-tracing/src/main/java/com/newrelic/SpanEventSender.java

```diff
@@ -0,0 +1,138 @@
+import com.newrelic.agent.model.SpanEvent;
+import com.newrelic.api.agent.Logger;
+import com.newrelic.api.agent.MetricAggregator;
+import com.newrelic.trace.v1.V1;
+import io.grpc.stub.ClientCallStreamObserver;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+
+class SpanEventSender implements Runnable {
+
+    private final Logger logger;
+    private final BlockingQueue<SpanEvent> queue;
+    private final MetricAggregator aggregator;
+    private final ChannelManager channelManager;
+
+    SpanEventSender(InfiniteTracingConfig config, BlockingQueue<SpanEvent> queue, MetricAggregator aggregator, ChannelManager channelManager) {
+        this.logger = config.getLogger();
+        this.queue = queue;
+        this.aggregator = aggregator;
+        this.channelManager = channelManager;
+    }
+
+    @Override
+    public void run() {
+        logger.log(Level.FINE, "Initializing {0}", this.getClass().getSimpleName());
+        while (true) {
+            try {
+                pollAndWrite();
+            } catch (Throwable t) {
+                logger.log(Level.SEVERE, t, "A problem occurred and no further spans will be sent.");
+                return;
+            }
+        }
+    }
+
+    @VisibleForTesting
+    void pollAndWrite() {
+        // Get stream observer
+        ClientCallStreamObserver<V1.Span> observer = channelManager.getSpanObserver();
+
+        // Confirm the observer is ready
+        if (!awaitReadyObserver(observer)) {
+            return;
+        }
+
+        // Poll queue for span
+        SpanEvent span = pollSafely();
+        if (span == null) {
+            return;
+        }
+
+        // Convert span and write to observer
+        V1.Span convertedSpan = convert(span);
+        writeToObserver(observer, convertedSpan);
+    }
+
+    @VisibleForTesting
+    boolean awaitReadyObserver(ClientCallStreamObserver<V1.Span> observer) {
+        if (observer.isReady()) {
+            return true;
+        }
+        try {
+            logger.log(Level.FINE, "Waiting for gRPC span observer to be ready.");
+            aggregator.incrementCounter("Supportability/InfiniteTracing/NotReady");
+            Thread.sleep(250);
+        } catch (InterruptedException exception) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException("Thread interrupted while awaiting ready gRPC span observer.");
+        }
+        return false;
+    }
+
+    @VisibleForTesting
+    SpanEvent pollSafely() {
+        try {
+            return queue.poll(250, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException("Thread interrupted while polling for spans.");
+        }
+    }
+
+    @VisibleForTesting
+    static V1.Span convert(SpanEvent spanEvent) {
```

Comment:
I would prefer the span conversion process to remain in a separate class. I know we discussed it is a straight forward and simple conversion. I think it is separate enough of a responsibility to organize on its own.

Reply:
Sure I can do that, but before I do, let me just explain my reasoning for why I didn't do it initially. 

IMO, decomposition is about breaking large problems into smaller problems of singular concern, and a separate java class files isn't always the best way to accomplish that. They can reduce readability (i.e. having to look at 5 files to figure out what's happening is harder to undertstand than 5 functions in a single file). And they can encourage consumption outside of code outside its intended use (i.e. a private function in a class file sends a clear signal to maintainers that only the owning class should consume it while a separate class might more readily reused). They can help reduce clutter, but at 138 lines I don't think that this class is suffering from that. 

Anyway, that's my point of view in general, but in this case I don't think breaking it out is such a bad thing.

Reply:
Appreciate the insight and this will help me in the future! At times, I've gone through what seems a bit much of code indirection and agree with your point it is hard to read.